### ISSUE TITLE
LTX-2 Phase 1: Video VAE (encoder + decoder)

### ISSUE THREAD

**Issue by twalderman** [BarkadaBrew/zimage.swift#112]
## Overview

Port LTX-2's 3D Video VAE to Swift/MLX. This encodes/decodes video frames to/from 128-channel latent space with 32x spatial and 8x temporal compression.

**Design doc:** `docs/design/ltx2-swift-port.md`
**Python reference:** `Blaizzy/mlx-video` → `mlx_video/models/video_vae/`
**Branch:** Create from `main`, name `codex/ltx2-phase1-vae`

## Files to Create

```
Sources/ComfyBox/LTX2/
├── VAE/
│   ├── LTX2VAE.swift              # Top-level encode/decode with tiled support
│   ├── LTX2Encoder3D.swift        # 3D encoder: conv_in → down_blocks → mid_block → conv_out
│   ├── LTX2Decoder3D.swift        # 3D decoder: conv_in → mid_block → up_blocks → conv_out
│   ├── LTX2ResnetBlock3D.swift    # ResNet block: GroupNorm → CausalConv3d → GroupNorm → CausalConv3d + skip
│   ├── LTX2DownBlock3D.swift      # Encoder downsample: N resnet blocks + spatial downsample
│   ├── LTX2UpBlock3D.swift        # Decoder upsample: N resnet blocks + spatial upsample
│   ├── LTX2MidBlock3D.swift       # Middle block: resnet → attention → resnet
│   └── LTX2Patchify.swift         # Patchify/unpatchify: (B,128,F,H,W) ↔ (B, F*H*W, 128)
├── Common/
│   ├── LTX2ModelConfig.swift      # Dataclass config for all LTX-2 components
│   └── LTX2WeightLoader.swift     # Weight loading + key remapping from HuggingFace format
```

## Key Specs

- **Input:** RGB frames `(B, F, H, W, 3)` where F must be `1 + 8k` (e.g., 1, 9, 17, 25...)
- **Output:** Latent `(B, 128, F/8, H/32, W/32)`
- **Compression:** 32x spatial (patch_size=4 × 3 stages of 2x downsample), 8x temporal
- **Latent channels:** 128 (vs 16 in SeedVR2)
- **Scaling factor:** Read from config (default ~0.3611 for LTX-2)
- **Convolutions:** CausalConv3d throughout — asymmetric temporal padding (causal)

## Reuse from SeedVR2

**Direct reuse:**
- `Upscale/SeedVR2/Common/CausalConv3d.swift` — verify padding matches LTX-2 (frame replication mode)
- `Upscale/SeedVR2/Common/SeedVR2RMSNorm.swift` — all normalization layers

**Adapt:**
- `Upscale/SeedVR2/VAE/SeedVR2Encoder3D.swift` → template for LTX2Encoder3D (different channel config: 128 latent channels)
- `Upscale/SeedVR2/VAE/SeedVR2Decoder3D.swift` → template for LTX2Decoder3D
- `Upscale/SeedVR2/VAE/SeedVR2ResnetBlock3D.swift` → template for LTX2ResnetBlock3D
- VAE tiling from SeedVR2 (64×64 latent tiles, stride 56, cosine ramp blending)

## Python Reference Details

**Encoder architecture** (from `video_vae/encoder.py`):
- `conv_in`: CausalConv3d(3, base_channels, kernel=3)
- `down_blocks`: List of DownBlock3D, each with N resnet blocks + optional downsample
- `mid_block`: MidBlock3D (resnet + attention + resnet)
- `conv_out`: CausalConv3d(last_channels, 2*latent_channels, kernel=3) — 2x for mean+logvar
- PixelNorm applied to encoder output

**Decoder architecture** (from `video_vae/decoder.py`):
- `conv_in`: CausalConv3d(latent_channels, last_channels, kernel=3)
- `mid_block`: MidBlock3D
- `up_blocks`: List of UpBlock3D (reverse of encoder), each with N resnet blocks + optional upsample
- `conv_out`: CausalConv3d(base_channels, 3, kernel=3)

**Weight key sanitization** (from `convert.py`):
- `model.diffusion_model.vae.encoder.` → strip prefix
- `model.diffusion_model.vae.decoder.` → strip prefix
- CausalConv3d weights: transpose from PyTorch `(O, I, D, H, W)` to MLX `(O, D, H, W, I)`

## Validation

1. Load VAE weights from `prince-canuma/LTX-2.3-distilled` (or download from `Lightricks/LTX-2.3`)
2. Encode a 256×256 single-frame test image
3. Decode the latent back to RGB
4. Compare with Python reference output (tolerance < 1e-4 for BF16)
5. Test tiled decode at 512×512 and 1024×1024

## Acceptance Criteria

- [ ] `LTX2VAE.encode()` produces correct latent shape and values
- [ ] `LTX2VAE.decode()` reconstructs frames from latents
- [ ] Tiled decode works for memory efficiency
- [ ] Weight loading handles both monolithic and modular weight formats
- [ ] `swift build -c release` passes with no warnings

**Comment by twalderman** [comment]
## Live debugging findings tonight (I2V, LTXV 2.3, real weights)

Ran the local I2V pipeline end-to-end for the first time with a complete weights set (`dgrauet/ltx-2.3-mlx`: transformer-distilled.safetensors, connector.safetensors, vae_encoder/decoder.safetensors + `unsloth/gemma-3-12b-it` text encoder). Two real bugs found and fixed along the way (see commits), but a third, unfixed issue remains and looks VAE-specific:

**Symptom:** Every I2V render (with or without a LoRA — isolated and confirmed identical artifact either way) produces output with a **uniform dot/mesh grid texture overlaying the whole frame**, plus washed-out/desaturated color, while the underlying **composition is correct** — you can clearly make out the right subject count/pose/layout matching the source image once you look past the grid overlay. That combination (correct high-level structure, uniform fine-grained texture corruption) is a classic signature of a **patch-reassembly ("unpatchify"/space-to-depth un-shuffle) bug in the VAE decoder** — the spatial compression un-shuffle is reassembling pixel patches in the wrong order/stride, not a random-weights or garbage-conditioning problem (those look categorically different — totally unrelated scene content, which is what we saw *before* the connector fix, not what we see now).

Ruled out:
- **Not the LoRA** — identical artifact with LoRAs completely removed from the request.
- **Not missing/garbage text conditioning** — before fixing the connector-weights bug (below), output was completely unrelated to both prompt and image (e.g. rendered a cartoon of children in front of a store sign). After the fix, output composition tracks the prompt/image correctly — this rules out the connector as the remaining cause.
- **Probably not gross VAE weight-loading omissions** — `LTX2WeightLoader.loadVAEWeightsFromTensors` logged full key coverage (encoder 86/86, decoder 86/86 weight tensors) during these runs, so it's not simply missing weights the way the connector was.

Next step for whoever picks this up: compare `LTX2VAE`'s decoder patchify/unpatchify (space-to-depth un-shuffle) implementation against the Python reference LTX-2.3 VAE decoder, specifically the ordering of the channel/spatial un-shuffle right before/after the final conv — that's the most likely site for a "right data, wrong pixel order" bug matching this exact symptom.

## Two related bugs found + fixed in the same session (commits incoming)

1. **`connector.safetensors` was silently allowed to be missing.** `LTX2TextEncoder.loadProjectionWeights()` fell back to *uninitialized* connector weights with only a `print()` warning when the file wasn't found, instead of throwing — producing a "successful" 200 response with completely garbage, prompt/image-uncorrelated output. Now throws `LTX2VideoError.weightsMissing(...)` instead. Also: the `--ltx2-weights` auto-download file-pattern list in `WarmServer.swift` didn't include `connector.safetensors` at all, so a from-scratch download would always hit this failure mode. Fixed in both places.

2. **`LTX2VideoCondition.strength`'s doc comment contradicts its own implementation.** The property doc says `/// Denoising strength (1.0 = full denoise, 0.0 = keep original)`, but `LTX2Conditioning.applyConditioning` computes `mask = 1.0 - strength`, and the mask-blend docstring right below it says `mask=0.0` "keeps the clean latent" / `mask=1.0` "uses the denoised output" — i.e. the *actual* behavior is **`strength=1.0` → mask=0.0 → frame kept as the original/clean image**, the opposite of the outer doc comment. Worth fixing the comment (or possibly the default — every request without an explicit `strength` gets 1.0, which per the *correct* reading should mean "frame 0 = exact source image," but that's not what we observed even before ruling in the VAE bug above — needs verification once the VAE issue is fixed, since right now it's hard to tell how much of the frame-0-not-matching-source symptom is the doc/semantics confusion vs. the VAE bug vs. both).

Environment: M3 Max, 128GB unified memory. Also hit an `OS_REASON_JETSAM` kill once when an image-generation request (krea2) came in on the shared queue while the LTX-2 pipeline (Gemma-3 12B + 38GB transformer + 6GB connector) was still warm in memory — worth keeping in mind for capacity planning if LTX-2 stays resident alongside image models rather than being evicted between uses.

**Comment by twalderman** [comment]
## Update: tiled decode does NOT fix it — ruling that out

Compared against the reference ComfyUI-LTXVideo workflow (downloaded and inspected `RuneXX/LTX-2.3-Workflows`' \`I2V_T2V_Simple_single_pass.json\` — the canonical single-pass I2V graph). Found the reference always decodes through \`VAEDecodeTiled\` (tile=512px, overlap=64px, temporal=4096, temporal_overlap=8), never a plain single-pass decode. ComfyBox's `LTX2VAE.decodeTiled()` already exists (cosine-ramp blended tiling, same pattern as the working SeedVR2 port) but was never wired up — `LTX2VideoGenerator.swift` constructs `LTX2PipelineConfig` without ever setting `tiledDecode`, so it silently defaulted to `false` (plain decode) for every render.

**Enabled it (`tiledDecode: true`) and re-tested — output is visually identical to the non-tiled run, same grid/mesh artifact.** So this was a real gap worth fixing for parity with the reference pipeline, but it is **not** the cause of the artifact. Ruling it out so nobody re-tries this path.

Also checked the reference's `LTXVImgToVideoInplace` node source (`comfy_extras/nodes_lt.py` in mainline ComfyUI) directly — its noise-mask formula (`mask = 1.0 - strength`) matches ComfyBox's `LTX2Conditioning.applyConditioning` exactly, so that part of the port is correct. Reference also confirmed no external pixel normalization before `vae.encode()` at the call site — ComfyBox's `normalizeForEncoder()` ([0,1]→[-1,1]) is still expected since the encoder module itself doesn't normalize internally (checked `LTX2VAE.encode()` — passes straight through to `encoder(input)`), so this isn't a double-normalization bug either.

**What's ruled out now:** LoRA application, connector/text conditioning (after the earlier fix), the noise-mask math, tiled-vs-non-tiled decode, external image normalization.

**What's still unexplained:** every render (regardless of the above) produces the same uniform dot/mesh grid texture with a curiously consistent artifact pattern (including what looks like a fixed text-like strip near the bottom) — that specific consistency across totally different prompts/seeds/strengths suggests something closer to a **degenerate/near-constant transformer output** (bad attention mask, RoPE, or a silently-mismapped weight in the DiT itself — not the VAE) than a decode-side bug. The 86/86-tensor-count VAE weight load also doesn't rule out a wrong-key-mapping bug (right count, wrong slots) — worth checking `LTX2WeightLoader`'s key remapping against the reference key names explicitly, not just checking counts match.

Next real step is dumping intermediate latents (post-transformer, pre-VAE) and diffing against a Python reference run on the same seed/prompt/image — that's the only way to actually localize this now; further guess-and-check against end-to-end video output isn't converging.

One reference resource worth downloading for a from-scratch comparison: `LiconStudio/Ltx2.3-VBVR-lora-I2V` and the various workflow collections at ltxworkflow.com / civitai (PlagueKind's "LTX23 workflow" collection) all document known-good I2V settings and node graphs for LTXV 2.3 — useful ground truth for whoever picks this up.

**Comment by twalderman** [comment]
## Root cause found & fixed: VAE encoder ran with zero-weight (dead) convolutions

**Commit:** `b4bc479` (pushed to `main`).

### Symptom
Local LTX-2 I2V produced a uniform grid/mesh on every frame AND ignored the source image (looked like T2V from the prompt).

### Decisive isolation
- **VAE decoder is fine.** Fed the *exact* Python-reference latent into the Swift decoder → smooth gradient matching the reference within bf16 noise (‑0.108 vs ‑0.104). No grid. So the grid is NOT a decode/unpatchify bug (patchify/unpatchify channel orderings were verified to match the reference `unpatchify_spatial`; dims 448×704 / 97 frames conform to ÷32 & 8n+1).
- **VAE encoder was broken.** Encoding a source image, the Swift latent had **per-channel spatial std = 0.0000 for all 128 channels** (only channel-to-channel variation, global std 0.21 vs reference 0.78). The latent was spatially constant → decodes to a tiled grid, and carries none of the image → T2V-like.
- Instrumented the encoder forward: spatial variance collapsed to 0 right at `conv_in` and again at `conv_out` (the resblocks only *looked* alive because their residual/shortcut paths carried spatial info around the dead convs; SpaceToDepthDownsample was fine because it already used `LTX2ConvWrapper`).

### Mechanism
The LTX‑2.3 `vae_encoder.safetensors` stores conv weights under a **nested `.conv.` path** (`conv_in.conv.weight`, `res_blocks.N.conv1.conv.weight`) — same as the decoder. But `LTX2Encoder3D` (conv_in/conv_out) and `LTX2ResnetBlock3D` (conv1/conv2) declared **bare `CausalConv3d`** modules expecting `conv_in.weight` (the old LTX‑2.0 0.7B layout). The loader's `update(parameters:, verify: [])` **silently dropped** the mismatched keys while still counting them → misleading "88/88 loaded". Every encoder conv_in/conv_out/resblock conv therefore ran with zero-init weights.

### Fix
Wrap the encoder's `conv_in`, `conv_out`, `conv1`, `conv2` in `LTX2ConvWrapper` (transparent passthrough → produces the nested `.conv.weight` key), exactly as the decoder already does. Forward math unchanged.

### Verification
- Encoder latent now matches Python reference: per-channel spatial std **0.766 vs 0.732**, global std **0.82 vs 0.78**.
- VAE round-trip of a source image is now **pixel-faithful** (no grid).
- Full I2V run: frame 0 now shows the actual source scene (orchard) instead of a uniform mesh — image is followed.

### Remaining (separate, secondary — not this bug)
Frame 0 is not yet pixel-identical (residual faint grid + slight purple tint + a horizontal seam), and the clip degrades to uniform gray by the last frame. These point to the transformer/sampler output quality and/or the multi-frame temporal decode (non-causal temporal convs bleed the degrading later frames into frame 0), plus the still-missing `LTXVPreprocess` CRF=33 H.264 round-trip on the conditioning image (out-of-distribution pristine PNG). Tracked separately — the primary dead-encoder bug is resolved.

**Comment by twalderman** [comment]
Shipped. LTX-2 Video VAE ported: Sources/ZImage/LTX2/VAE/{LTX2VAE,LTX2Encoder3D,LTX2Decoder3D,LTX2ResnetBlock3D,LTX2Patchify}.swift, commit be2730b "feat(ltx2): Phase 1 — Video VAE encoder/decoder".

— triage 2026-09-04 (Claude)
